app/models/template.rb

```diff
@@ -53,24 +53,25 @@ def default
       unarchived.where(is_default: true, published: true).order(:version).last
     end
   end
-
-  # Creates a copy of the current template
-  # raises ActiveRecord::RecordInvalid when save option is true and validations fails
-  def deep_copy(**options)
-    copy = self.dup
-    copy.version = options.fetch(:version, self.version)
-    copy.published = options.fetch(:published, self.published)
-    copy.save! if options.fetch(:save, false)
-    self.phases.each{ |phase| copy.phases << phase.deep_copy(options) }
-    return copy
-  end
-
+  
```

Comment:
I'm not convinced to privatise this method, it is pretty flexible thanks to its configuration options and therefore it is safe the leave it as public since its defaults are sensible, i.e. it does not save a hierarchy. Moreover, can help to create specific unit tests without having to use generate_copy, customize, upgrade_customization... At the end in ruby even private methods are accessible (e.g. using send)